kernel_gateway/services/notebooks/handlers.py

```diff
@@ -29,82 +34,86 @@ def _request_assignment_for_lang(self, kernel_name, expression):
-            return False, execute_result
-        else:
-            return False, ''.join(stream_messages)
-
-    def _send_code(self, code):
-        # execute the code and return the result and HTTP status code
-        parent_header = self.kernel_client.execute(code)
-        status = 200
-        error, result = self._process_messages(parent_header=parent_header)
-
-        if error:
-            status = 500
-
-        return result, status
-
+    def on_recv(self, msg):
+        # Only look at messages which are derived from the parent_header
+        # TODO Refactor this so we only look for parent headers for the actual cell execution
+        if msg['parent_header']['msg_id'] == self.parent_header:
+            # On idle status, exit our loop
+            if msg['header']['msg_type'] == 'status' and msg['content']['execution_state'] == 'idle':
+                result = {'status': 200, 'content': ''}
+                if self.error_message:
+                    result['content'] =self.error_message
+                    result['status'] = 500
+                elif self.execute_result and self.execute_result is not '':
+                    result['content'] = self.error_message
+                else:
+                    result['content'] = ''.join(self.stream_messages)
+                self.execution_future.set_result(result)
+            # Store the execute result
+            elif msg['header']['msg_type'] == 'execute_result':
+                self.execute_result = msg['content']['data']
+            # Accumulate the stream messages
+            elif msg['header']['msg_type'] == 'stream':
+                self.stream_messages.append(msg['content']['text'])
+            # Store the error message
+            elif msg['header']['msg_type'] == 'error':
+                self.error_message = 'Error {}: {} \n'.format(
+                    msg['content']['ename'],
+                    msg['content']['evalue']
+                )
+
+    @gen.coroutine
     def _handle_request(self):
+        self.execute_result = None
+        self.stream_messages = []
+        self.error_message = None
         if self.request.method not in self.sources:
             self.set_status(405)
             self.finish()
             return
 
-        source_code = self.sources[self.request.method]
-        REQUEST = json.dumps({
-            'body' : parse_body(self.request.body),
-            'args' : parse_args(self.request.arguments),
-            'path' : self.path_kwargs
-        })
-        request_code = self._request_assignment_for_lang(self.kernel_name, REQUEST)
-
-        access_log.debug('Request code for notebook cell is: {}'.format(request_code))
-        self._send_code(request_code)
-        result, status = self._send_code(source_code)
+        self.execution_future = Future()
+        self.response_future = Future()
+        kernel_client, kernel_id = yield self.kernel_pool.acquire()
+        self.kernel_pool.on_recv(kernel_id, self.on_recv)
```

Comment:
Does this on_recv line need to be in the try also? if it throws an exception, does that mean the kernel is never released back into the pool?